src/workerd/api/modules.c++

```diff
@@ -36,4 +34,50 @@ jsg::JsRef<jsg::JsValue> EnvModule::withEnv(
   });
 }
 
+kj::Maybe<jsg::JsObject> EnvModule::getCurrentExports(jsg::Lock& js) {
+  auto& key = jsg::IsolateBase::from(js.v8Isolate).getExportsAsyncContextKey();
+  // Check async context first - withExports() overrides take precedence over the disable flags.
+  KJ_IF_SOME(frame, jsg::AsyncContextFrame::current(js)) {
+    KJ_IF_SOME(value, frame.get(key)) {
+      auto handle = value.getHandle(js);
+      if (handle->IsObject()) {
+        return jsg::JsObject(handle.As<v8::Object>());
+      }
+    }
+  }
+  // Either disable flag prevents base exports access.
+  if (!FeatureFlags::get(js).getEnableImportableExports() ||
+      FeatureFlags::get(js).getDisableImportableEnv()) {
+    return kj::none;
+  }
+
+  return js.getWorkerExports().map([&](const jsg::V8Ref<v8::Object>& val) -> jsg::JsObject {
+    return jsg::JsObject(val.getHandle(js));
+  });
+}
+
+jsg::JsRef<jsg::JsValue> EnvModule::withExports(
+    jsg::Lock& js, jsg::Value newExports, jsg::Function<jsg::JsRef<jsg::JsValue>()> fn) {
+  auto& key = jsg::IsolateBase::from(js.v8Isolate).getExportsAsyncContextKey();
+  jsg::AsyncContextFrame::StorageScope storage(js, key, kj::mv(newExports));
+  return js.tryCatch([&]() mutable -> jsg::JsRef<jsg::JsValue> { return fn(js); },
+      [&](jsg::Value&& exception) mutable -> jsg::JsRef<jsg::JsValue> {
+    js.throwException(kj::mv(exception));
+  });
+}
+
+jsg::JsRef<jsg::JsValue> EnvModule::withEnvAndExports(jsg::Lock& js,
+    jsg::Value newEnv,
+    jsg::Value newExports,
+    jsg::Function<jsg::JsRef<jsg::JsValue>()> fn) {
+  auto& envKey = jsg::IsolateBase::from(js.v8Isolate).getEnvAsyncContextKey();
+  auto& exportsKey = jsg::IsolateBase::from(js.v8Isolate).getExportsAsyncContextKey();
```

Comment:
I wonder if it's worth merging the two keys into one key that holds both values?

But @jasnell would know better than me, so I'll defer to his opinion.